src/extensions/default/CodeFolding/main.js

```diff
@@ -0,0 +1,352 @@
+      *         if (debug) {
+      *             logMessages.forEach(function (m) {
+      *                 console.debug(m);
+      *             });
+      *         }
+      *     }
+      */
+    function collapseAll() {
+        var editor = EditorManager.getFocusedEditor();
+        if (editor && editor._codeMirror) {
+            var cm = editor._codeMirror;
+            CodeMirror.commands.foldToLevel(cm);
+        }
+    }
+
+    /**
+      * Expands all folded regions in the current document
+      */
+    function expandAll() {
+        var editor = EditorManager.getFocusedEditor();
+        if (editor && editor._codeMirror) {
+            var cm = editor._codeMirror;
+            CodeMirror.commands.unfoldAll(cm);
+        }
+    }
+
+    /**
+      * Initialises and creates the code-folding gutter.
+      * @param {Editor} editor the editor on which to initialise the fold gutter
+      */
+    function createGutter(editor) {
+        var cm = editor._codeMirror;
+        if (!cm) {
+            return;
+        }
+        var path = editor.document.file.fullPath, _lineFolds = prefs.getFolds(path);
+        _lineFolds = _lineFolds || {};
+        cm._lineFolds = _lineFolds;
+        var gutters = cm.getOption("gutters").slice(0);
+        var lnIndex = gutters.indexOf("CodeMirror-linenumbers");
+        gutters.splice(lnIndex + 1, 0, gutterName);
+        cm.setOption("gutters",  gutters);
+        cm.setOption("foldGutter", {onGutterClick: onGutterClick});
+
+        $(cm.getGutterElement()).on({
+            mouseenter: function () {
+                if (prefs.getSetting("fadeFoldButtons")) {
+                    foldGutter.updateInViewport(cm);
+                }
+            },
+            mouseleave: function () {
+                if (prefs.getSetting("fadeFoldButtons")) {
+                    foldGutter.clearGutter(cm);
+                }
+            }
+        });
+    }
+
+    /**
+      * Event handler to initialise fold-gutter and restores/saves line folds in editors whenever the active editor changes
+      * @param {object} event the event object
+      * @param {Editor} current the current editor
+      * @param {Editor} previous the previous editor
+      */
+    function onActiveEditorChanged(event, current, previous) {
+        if (prefs.getSetting("enabled")) {
+            if (current && current._codeMirror.getOption("gutters").indexOf(gutterName) === -1) {
+                createGutter(current);
+                restoreLineFolds(current);
+            }
+            if (previous) {
+                saveLineFolds(previous);
+            }
+        }
+    }
+
+    /**
+      * Saves the line folds in the current full editor before it is closed.
+      */
+    function saveBeforeClose() {
+        saveLineFolds(EditorManager.getCurrentFullEditor());
+    }
+
+    /**
+        Initialise the extension
+    */
+    function init() {
+        if (CodeMirror.fold.combine || !prefs.getSetting("enabled")) {
```

Comment:
It would be truthy if the code-folding plugin is already loaded. This is mainly  for the extension in the registry- to prevent registering key bindings twice etc. I could not find an api to check that an extension has loaded. That code fragment can be removed without affecting the logic of initialisation.